### PR TITLE
feat: OpenAI compatible API

### PR DESCRIPTION
**Pull request by XunchaoZ**
The PR adds support for OpenLLM to be compatible with the legacy completions API of OpenAI for both stream & non-stream responses. 

Notes:
1. The 'usage' (# of tokens) and 'finish_reason' are not currently supported due to the fact that a lot of implementations in _llm.py only take the 'text' of the output as the response. Changing them would raise inconsistency with other API endpoints, so I will have a discussion on how to proceed on the next step. 
2. There is inconsistency in the implementation of vllm with pytorch. Probably need to discuss with Aaron on how to make everything consistent. With his guidance, I can make those changes.
3. Still need to do the chat completion API and embedding API, but should be much easier after completing this.

### CODE REVIEW

File: openllm-python/src/openllm/_llm.py [hunk 1]
```diff
@@ -738,7 +738,8 @@ def postprocess_generate(self, prompt: str, generation_result: t.Any, **attrs: t
     > [!NOTE]
     > This will be used from the client side.
     '''
-    if isinstance(generation_result, dict) and 'text' in generation_result: return generation_result['text']
+    import pdb; pdb.set_trace()
```
Comment (aarnphm) [inline openllm-python/src/openllm/_llm.py:741]:
```suggestion
```

File: openllm-python/src/openllm/_llm.py [hunk 2]
```diff
@@ -988,14 +988,16 @@ def generate_iterator(self,
       past_key_values = out = token = None
       finish_reason = None
       for i in range(config['max_new_tokens']):
-        torch.cuda.synchronize()
+        if torch.cuda.is_available():
+          torch.cuda.synchronize()
```
Comment (aarnphm) [inline openllm-python/src/openllm/_llm.py:992]: You can make this one line

and lets make a separate PR for this. This doesn't seem related at all.

File: openllm-python/src/openllm/_service.py [hunk 3]
```diff
@@ -69,6 +71,90 @@ async def generate_stream_v1(input_dict: dict[str, t.Any]) -> t.AsyncGenerator[s
   else:
     return runner.generate_iterator.async_stream(qa_inputs.prompt, adapter_name=qa_inputs.adapter_name, echo=echo, **qa_inputs.llm_config.model_dump())
 
+@svc.api(route='v1/completions', input=_JsonInput, output=bentoml.io.Text())
+async def completion_v1(input_dict: dict[str, t.Any], ctx: bentoml.Context):
+  stream = input_dict.pop('stream', False)        # Determine whether to stream the response
```
Comment (aarnphm) [inline openllm-python/src/openllm/_service.py:76]: are you using tab here? I think it should just be two spaces.

File: openllm-python/src/openllm/_service.py [hunk 4]
```diff
@@ -69,6 +71,90 @@ async def generate_stream_v1(input_dict: dict[str, t.Any]) -> t.AsyncGenerator[s
...
+  default_config = llm_config.model_dump(flatten=True)
+
+  for key in list(input_dict.keys()):
+    if key == 'prompt': continue
+    elif key == 'max_tokens':
+      default_config['max_new_tokens'] = input_dict[key]
+    elif key in default_config:
+      default_config[key] = input_dict[key]   # Replace the default config with the input config
+    del input_dict[key]                       # Remove the key from input_dict
+  input_dict['adapter_name'] = None
```
Comment (aarnphm) [inline openllm-python/src/openllm/_service.py:88]: I think we shouldn't do this. Probably best to have a function on the `LLMConfig` to convert to openai attributes or something.

File: openllm-python/src/openllm/_llm.py (hunk 2 above)
Comment (XunchaoZ) [reply openllm-python/src/openllm/_llm.py:992]: OK no problem.

File: openllm-python/src/openllm/_service.py (hunk 3 above)
Comment (XunchaoZ) [reply openllm-python/src/openllm/_service.py:76]: Yes, sure I'll do that.

File: openllm-python/src/openllm/_service.py (hunk 4 above)
Comment (XunchaoZ) [reply openllm-python/src/openllm/_service.py:88]: Yes that's right, we could do that.

File: openllm-python/src/openllm/conversion/__init__.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from __future__ import annotations
+import typing as t
+
+def openllm_to_openai_config(input_dict: dict[str, t.Any], default_config: dict[str, t.Any]) -> dict[str, t.Any]:
```
Comment (aarnphm) [inline openllm-python/src/openllm/conversion/__init__.py:4]: Let's just move this file into `protocol/utils.py`, or you should make this a function of `LLMConfig`. This can be found under `openllm_core`.

```python

openai_config = openllm.AutoConfig.for_model('llama').to_openai_config()

File: openllm-python/src/openllm/protocol/openai.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
+from __future__ import annotations
+import time
+
+import attr
+
+import openllm_core
+
+@attr.define
+class LogProbs:
+    text_offset: list[int] = attr.ib(default=attr.Factory(list))
```
Comment (aarnphm) [inline openllm-python/src/openllm/protocol/openai.py:10]: Let's use `attr.field` instead of ib.

Comment (aarnphm) [reply openllm-python/src/openllm/protocol/openai.py:10]: Note that since we are compat with 3.8, the type hint will fail.

The current type annotations that you have is according to [PEP604](https://peps.python.org/pep-0604/), which we are currently using. But for attrs class that does serialisation. Best to do just do 

```python

import typing as t

@attr.define
class LogProbs:
  text_offset: t.List[int] = ...
```

File: openllm-python/src/openllm/_service.py [hunk 7]
```diff
@@ -69,6 +74,48 @@ async def generate_stream_v1(input_dict: dict[str, t.Any]) -> t.AsyncGenerator[s
...
+  echo = input_dict.pop('echo', False)
+  default_config = llm_config.model_dump(flatten=True)
+  input_dict = conversion.openllm_to_openai_config(input_dict, default_config)
+  qa_inputs = openllm.GenerationInput.from_llm_config(llm_config)(**input_dict)
+  config = qa_inputs.llm_config.model_dump()
+
+  def completion_response(responses: list[str]):
+    choices = [openai.CompletionTextChoice(text=response, index=i) for i, response in enumerate(responses)]
+    result = openai.CompletionResponse(choices=choices, model=model) # TODO: logprobs, finish_reason and usage
+    print(attr.asdict(result))
```
Comment (aarnphm) [inline openllm-python/src/openllm/_service.py:89]:
```suggestion
```

File: openllm-python/src/openllm/_service.py [hunk 8]
```diff
@@ -1,9 +1,11 @@
 # mypy: disable-error-code="call-arg,misc,attr-defined,type-abstract,type-arg,valid-type,arg-type"
 from __future__ import annotations
+import json
```
Comment (aarnphm) [inline openllm-python/src/openllm/_service.py:3]: no need to use json. we are using orjson for serde.

**Review by aarnphm — CHANGES_REQUESTED** [review]
left a few comments

File: openllm-python/src/openllm/_service.py [hunk 9]
```diff
@@ -69,6 +69,88 @@ async def generate_stream_v1(input_dict: dict[str, t.Any]) -> t.AsyncGenerator[s
   else:
     return runner.generate_iterator.async_stream(qa_inputs.prompt, adapter_name=qa_inputs.adapter_name, echo=echo, **qa_inputs.llm_config.model_dump())
 
+@svc.api(route='v1/completions', input=_JsonInput, output=bentoml.io.Text())
+async def completion_v1(input_dict: dict[str, t.Any], ctx: bentoml.Context) -> str | t.AsyncGenerator[str, None]:
+  stream = input_dict.pop('stream', False)  # Determine whether to stream the response
+  echo = input_dict.pop('echo', False)
+  config = llm_config.to_openai_config(input_dict)
+  qa_inputs = openllm.GenerationInput.from_llm_config(llm_config)(**config)
+  config = qa_inputs.llm_config.model_dump()
```
Comment (aarnphm) [inline openllm-python/src/openllm/_service.py:78]: btw this logic seems wrong. config at L76 will be overwritten by L78

File: openllm-core/src/openllm_core/_configuration.py [hunk 10]
```diff
@@ -1356,6 +1356,24 @@ def model_dump(self, flatten: bool = False, **_: t.Any) -> DictStrAny:
   def model_dump_json(self, **kwargs: t.Any) -> bytes:
     return orjson.dumps(self.model_dump(**kwargs))
 
+  def to_openai_config(self, new_config: t.Any) -> t.Any:
+    default_config = bentoml_cattr.unstructure(self)
+    generation_config = bentoml_cattr.unstructure(self.generation_config)
+    sampling_config = bentoml_cattr.unstructure(self.sampling_config)
+    default_config.update(generation_config)
+    default_config.update(sampling_config)
```
Comment (aarnphm) [inline openllm-core/src/openllm_core/_configuration.py:1364]: what is the reason for doing this? shouldn't `to_openai_config` just get all of the available options that OpenAI API accepts?

All of the currently attributes for LLMConfig can still be accessed within the service definition.

File: openllm-python/src/openllm/protocol/openai.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+  choices: t.List[ChatCompletionStreamChoice]
+  model: str
+  object: str = 'chat.completion.chunk'
+  id: str = attr.field(default=attr.Factory(lambda: openllm_core.utils.gen_random_uuid('chatcmpl')))
+  created: int = attr.field(default=attr.Factory(lambda: int(time.time())))
+
+def openai_messages_to_openllm_prompt(messages: list[Message]) -> str:
+  # TODO: Improve the prompt
+  formatted = '\n'.join([f"{message['role']}: {message['content']}" for message in messages])
+  return f"Complete the assistant's response. Use system info if provided.\n{formatted}\nassistant:"
```
Comment (aarnphm) [inline openllm-python/src/openllm/protocol/openai.py:89]: Is this the prompt that all OpenAI's compatible API use?

**Review by aarnphm — CHANGES_REQUESTED** [review]

File: openllm-python/src/openllm/protocol/openai.py [hunk 12]
```diff
@@ -8,16 +8,16 @@
...
+  token_logprobs: t.List[float] = attr.field(default=attr.Factory(list))
+  tokens: t.List[str] = attr.field(default=attr.Factory(list))
+  top_logprobs: t.List[t.Dict[str, t.Any]] = attr.field(default=attr.Factory(list))
 
 @attr.define
 class CompletionTextChoice:
   text: str
   index: int
-  logprobs: LogProbs = attr.field(default=None)
+  logprobs: LogProbs = attr.field(default=attr.Factory(lambda: LogProbs()))
```
Comment (aarnphm) [inline openllm-python/src/openllm/protocol/openai.py]: we don't have logprob anw right? so probably default to None should be fine?

File: openllm-python/src/openllm/_service.py [hunk 13]
```diff
@@ -115,7 +115,7 @@ async def chat_completion_v1(input_dict: dict[str, t.Any], ctx: bentoml.Context)
   del input_dict['messages']
 
   config = llm_config.to_openai_config(input_dict)
-  qa_inputs = openllm.GenerationInput.from_llm_config(llm_config)(**input_dict)
+  qa_inputs = openllm.GenerationInput.from_llm_config(llm_config)(**config)
```
Comment (aarnphm) [inline openllm-python/src/openllm/_service.py:118]: Let's chat about this, I think the implementation is still incorrect.

**Review by aarnphm — CHANGES_REQUESTED** [review]

**Comment by aarnphm** [comment]
pre-commit.ci autofix

**Review by aarnphm — APPROVED** [review]
